### PR TITLE
Enhancement/2044 register site on proxy

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue:

- #2044

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Authentication/Google_Proxy.php [hunk 1]
```diff
@@ -330,6 +354,33 @@ public function get_site_fields() {
...
+			'mode'             => '',
+			'hl'               => $this->context->get_locale( 'user' ),
+			'application_name' => self::get_application_name(),
+			'service_version'  => Feature_Flags::enabled( 'serviceSetupV2' ) ? 'v2' : '',
+		);
+
+		/**
+		 * Filters the setup mode.
+		 *
+		 * @since n.e.x.t
```
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:377]: This is missing a `$param` doc for the filtered variable. In WordPress hooks you can just put any variable name there, e.g. `$mode`, it doesn't have to match the name of what's really passed to it.

File: includes/Core/Authentication/Google_Proxy.php [hunk 2]
```diff
@@ -399,24 +450,59 @@ public function unregister_site( Credentials $credentials ) {
...
-				'mode' => $mode,
-				'body' => $this->get_site_fields(),
+				'return' => 'response',
+				'mode'   => $mode,
+				'body'   => array_merge(
+					$this->get_site_fields(),
+					$this->get_user_fields(),
+					$this->get_metadata_fields(),
+					array(
+						'scopes' => implode( ' ', $this->required_scopes ),
```
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:486]: Similar to OAuth, this parameter has to be called `scope` (singular), even though it's multiple scopes.

File: includes/Core/Authentication/Google_Proxy.php [hunk 3]
```diff
@@ -399,24 +450,59 @@ public function unregister_site( Credentials $credentials ) {
...
+	 * @since n.e.x.t Updated the function to return redirect URL.
 	 *
 	 * @param Credentials $credentials Credentials instance.
 	 * @param string      $mode        Sync mode.
-	 * @return array|WP_Error Response of the wp_remote_post request.
+	 * @return string|WP_Error Redirect URL on success, otherwise an error.
 	 */
-	public function sync_site_fields( Credentials $credentials, $mode = 'async' ) {
-		return $this->request(
+	public function sync_site_fields( Credentials $credentials = null, $mode = 'async' ) {
```
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:474]: This method should _always_ be called with credentials being passed, to sync site fields they need to be present.

I like your idea of reusing the same logic, but could we do it in a way that both public methods "make sense" by themselves? How about we introduce a private method with the parts that is common between them which can be controlled with whatever parameters are needed, but the public method should "suggest" the right parameters that are needed.

Basically this method's parameters should remain like they were before, with `Credentials` instance being required.

File: includes/Core/Authentication/Google_Proxy.php [hunk 4]
```diff
@@ -399,24 +450,59 @@ public function unregister_site( Credentials $credentials ) {
...
 
+	/**
+	 * Registers the site on the proxy.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string|WP_Error Redirect URL on success, otherwise an error.
+	 */
+	public function register_site() {
+		return $this->sync_site_fields( null, null );
```
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:461]: See my other comment, maybe better to use a private "low-level" method here. Also, let's either also add the `$mode = 'async'` parameter so that it can be controlled from the outside, or we need to specify `sync` here because the only use-case we have for it now needs to be `sync`.

For parity with the other method, I prefer having a parameter for it that can be controlled from the outside.

File: includes/Core/Authentication/Setup_V2.php [hunk 5]
```diff
@@ -54,15 +54,23 @@ public function handle_action_setup_start() {
...
 		}
 
-		if ( $this->credentials->has() ) {
-			$this->google_proxy->sync_site_fields( $this->credentials, 'sync' );
+		$required_scopes = $this->authentication->get_oauth_client()->get_required_scopes();
+		$this->google_proxy->with_scopes( $required_scopes );
+
+		$oauth_setup_redirect = $this->credentials->has()
+			? $this->google_proxy->sync_site_fields( $this->credentials, 'sync' )
+			: $this->google_proxy->register_site();
```
Comment (felixarntz) [inline includes/Core/Authentication/Setup_V2.php:62]: Just pointing out here that, related to my comments above, we may need to provide `'sync'` here if we introduce a parameter for it - I think that would be more consistent.

File: includes/Modules/Analytics.php [hunk 6]
```diff
@@ -121,6 +121,13 @@ public function register() {
 		add_action( 'template_redirect', $this->get_method_proxy( 'register_tag' ) );
 
 		add_filter( 'googlesitekit_proxy_setup_url_params', $this->get_method_proxy( 'update_proxy_setup_mode' ) );
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:123]: This filter call (and also the entire `googlesitekit_proxy_setup_url_params` filter being applied as well) should be removed, per the ACs.

File: includes/Modules/Analytics.php [hunk 7]
```diff
@@ -1404,11 +1411,25 @@ public function handle_token_response_data( $token_response ) {
...
 	private function update_proxy_setup_mode( $params ) {
-		if ( Feature_Flags::enabled( 'serviceSetupV2' ) && ! $this->is_connected() ) {
-			$params['mode'] = 'analytics-step';
+		$mode = $this->get_proxy_setup_mode();
+		if ( ! empty( $mode ) ) {
+			$params['mode'] = $mode;
 		}
 
 		return $params;
 	}
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:1420]: See above, this method can then be removed.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php [hunk 8]
```diff
@@ -334,14 +334,22 @@ public function test_sync_site_fields() {
...
+				'scopes',
+				'service_version',
+				'site_id',
+				'site_secret',
+				'supports',
+				'url',
+				'user_roles',
 			),
 			array_keys( $this->request_args['body'] )
 		);
```
Comment (felixarntz) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:355]: While this would be somewhat similar, could you also add one simple test for the `register_site` method?

File: includes/Core/Authentication/Setup_V2.php [hunk 9]
```diff
@@ -54,15 +54,23 @@ public function handle_action_setup_start() {
...
+		if ( is_wp_error( $oauth_setup_redirect ) || ! filter_var( $oauth_setup_redirect, FILTER_VALIDATE_URL ) ) {
+			wp_die( esc_html__( 'The request to the authentication proxy has failed. Please, try again later.', 'google-site-kit' ) );
 		}
 
 		if ( $redirect_url ) {
 			$this->user_options->set( OAuth_Client::OPTION_REDIRECT_URL, $redirect_url );
 		}
 
-		$this->redirect_to_proxy();
+		wp_redirect( $oauth_setup_redirect ); // phpcs:ignore WordPress.Security.SafeRedirect.wp_redirect_wp_redirect
```
Comment (felixarntz) [inline includes/Core/Authentication/Setup_V2.php:72]: Can't we use `wp_safe_redirect` here?

**Review by felixarntz — COMMENTED** [review]
@eugene-manuilov Looks great, only a few points to review.

File: includes/Core/Authentication/Google_Proxy.php (hunk 1 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php:377]: Updated.

File: includes/Core/Authentication/Google_Proxy.php (hunk 2 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php]: Updated.

File: includes/Core/Authentication/Google_Proxy.php (hunk 3 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php]: Yep, makes sense. Updated.

File: includes/Core/Authentication/Google_Proxy.php (hunk 4 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php]: Updated.

File: includes/Core/Authentication/Setup_V2.php (hunk 5 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Setup_V2.php]: Updated.

File: includes/Modules/Analytics.php (hunk 6 above)
Comment (eugene-manuilov) [reply includes/Modules/Analytics.php]: Removed.

File: includes/Modules/Analytics.php (hunk 7 above)
Comment (eugene-manuilov) [reply includes/Modules/Analytics.php]: Removed.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php (hunk 8 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:386]: Added.

File: includes/Core/Authentication/Setup_V2.php (hunk 9 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Setup_V2.php:72]: No, because the redirect URL has a different domain.

File: includes/Modules/Analytics.php (hunk 6 above)
Comment (felixarntz) [reply includes/Modules/Analytics.php]: You've now removed the filter _usage_ here, can you also remove the entire filter itself please? The filter hook shouldn't even exist anymore.

**Review by felixarntz — COMMENTED** [review]
@eugene-manuilov New changes lgtm, but please review the two comment replies above. I resolved all other comments.

File: includes/Modules/Analytics.php (hunk 6 above)
Comment (eugene-manuilov) [reply includes/Modules/Analytics.php]: Ok, got it. Removed the filter too.

File: includes/Core/Authentication/Setup_V2.php (hunk 9 above)
Comment (eugene-manuilov) [reply includes/Core/Authentication/Setup_V2.php:72]: I don't know because I can't verify it on my local. When I try to set up the plugin with the V2 setup flow, I get the following error: 

```
We were unable to send credentials to your WordPress site. Please go back to your site and restart the setup. The following error was returned: No route was found matching the URL and request method.
```

Is it something wrong on the WP side or you need to adjust something on the proxy side?

File: includes/Core/Authentication/Google_Proxy.php [hunk 10]
```diff
@@ -330,6 +347,35 @@ public function get_site_fields() {
...
+	 * Gets metadata fields.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Metadata fields array.
+	 */
+	public function get_metadata_fields() {
+		$metadata = array(
+			'supports'         => implode( ',', $this->get_supports() ),
+			'nonce'            => wp_create_nonce( self::ACTION_SETUP ),
```
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:360]: This has to use `self::NONCE_ACTION` for the action per the changes in #4383 and #4376.

File: includes/Core/Authentication/Setup_V2.php [hunk 11]
```diff
@@ -54,15 +54,23 @@ public function handle_action_setup_start() {
...
+		exit;
 	}
```
Comment (felixarntz) [inline includes/Core/Authentication/Setup_V2.php:75]: There is another problem further down in this file that I discovered during testing, but it's more of an oversight from #4376 than it belongs to this issue. Still, let's fix it in here: In `Setup_V2::handle_action_exchange_site_code`, if the request is successful, we currently redirect back to the proxy without adding the `site_id` query parameter which causes an exception.

File: includes/Core/Authentication/Setup_V2.php (hunk 9 above)
Comment (felixarntz) [reply includes/Core/Authentication/Setup_V2.php:72]: I just tested this (note that you have to use the staging proxy) and found a few other small, but crucial problems that were preventing the flow from working. But yes, I then verified that `wp_safe_redirect` works here as expected, since we allowlist the host.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@eugene-manuilov Found a few minor, but critical problems during my testing (see below).

Since I've already addresed them locally, I'll push these updates here myself.

**Review by felixarntz — APPROVED** [review]
With the additional changes, lgtm!

Will assign this back to you @eugene-manuilov for a final sanity check on my additional changes. I've already added a QAB for this.

**Comment by eugene-manuilov** [comment]
> With the additional changes, lgtm!
> 
> Will assign this back to you @eugene-manuilov for a final sanity check on my additional changes. I've already added a QAB for this.

Thanks, @felixarntz. All looks good to me. I can't approve mine PR. So assigning this back to you.
